contracts/ColonyNetworkAuction.sol

```diff
@@ -0,0 +1,203 @@
+  // Total number of auctioned tokens
+  uint public quantity;
+  // CLNY Wei per Token wei
+  uint public finalPrice;
+  uint public constant MULTIPLIER = 10 ** 18;
+  bool public finalized;
+  
+  mapping (address => uint256) public bids;
+
+  modifier auctionNotFinalized() {
+    require(!finalized);
+    _;
+  }
+
+  modifier auctionFinalized {
+    require(finalized);
+    _;
+  }
+
+  modifier endBlockNotSet {
+    require(endBlock == 0);
+    _;
+  }
+
+  modifier endBlockSet {
+    require(endBlock > 0);
+    _;
+  }
+
+  modifier allBidsClaimed  {
+    require(tokensClaimed == receivedTotal);
+    _;
+  }
+
+  event AuctionStarted(uint indexed _startBlock);
+  event AuctionBid(address indexed _sender, uint _amount, uint _missingFunds);
+  event AuctionClaim(address indexed _recipient, uint _sentAmount);
+  event AuctionFinalized(uint _finalPrice);
+
+  function DutchAuction(
+    address _clnyToken,
+    address _token,
+    uint256 _quantity,
+    uint _startPrice) 
+    public 
+    {
+    colonyNetwork = msg.sender;
+    require(_clnyToken != 0x0 && _token != 0x0);
+    clnyToken = ERC20Extended(_clnyToken);
+    token = ERC20Extended(_token);
+    
+    require(_startPrice > 0);
+    startPrice = _startPrice;
+    require(_quantity > 0);
+    quantity = _quantity;
+    startBlock = block.number;
+    AuctionStarted(startBlock);
+  }
+
+  function bid(uint256 _amount) public
+  auctionNotFinalized
+  endBlockNotSet
+  {
+    require(_amount > 0);
+    uint _totalToEndAuction = totalToEndAuction();
+    uint remainingToEndAuction = sub(_totalToEndAuction, receivedTotal);
+
+    // Adjust the amount for final bid in case that takes us over the offered quantity at current price
+    // Also conditionally set the endBlock
+    uint amount;
+    if (remainingToEndAuction > _amount) {
+      amount = _amount;
+    } else {
+      amount = remainingToEndAuction;
+      endBlock = block.number;
+    }
+    
+    clnyToken.transferFrom(msg.sender, this, amount);
+    bids[msg.sender] = add(bids[msg.sender], amount);
+    receivedTotal = add(receivedTotal, amount);
+    
+    AuctionBid(msg.sender, amount, sub(remainingToEndAuction, amount));
+  }
+
+  // Finalize the auction and sets the final Token price
+  function finalize() public
+  auctionNotFinalized
+  endBlockSet
+  {
+    // Give the network all CLNY sent to the auction in bids
+    clnyToken.transfer(colonyNetwork, receivedTotal);
+    finalPrice = mul(receivedTotal, MULTIPLIER) / quantity;
+    finalized = true;
+    AuctionFinalized(finalPrice);
+  }
+
+  function claim() public 
+  auctionFinalized
+  returns (bool)
+  {
+    uint amount = bids[msg.sender];
+    uint tokens = mul(amount, MULTIPLIER) / finalPrice;
+
+    // Due to finalPrice floor rounding, the number of assigned tokens may be higher
+    // than expected. Therefore, the number of remaining unassigned auction tokens
+    // may be smaller than the number of tokens needed for the last claimTokens call
```

Comment:
This is unfair to the last claimant, though I agree only a little bit. Instead, when finalizing the auction, we could add `1` to the price after the division, and keep track of the number of people that have claimed, rather than the number of tokens that have been claimed. Leftovers could be returned to the `ColonyNetwork`.